plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php

```diff
@@ -0,0 +1,225 @@
+	 * @throws \ReflectionException Something went wrong when using reflection to get information about the class to resolve.
+	 */
+	private function instantiate_class_using_reflection( string $class_name, array &$resolve_chain ): object {
+		$ref_class = new \ReflectionClass( $class_name );
+		$instance  = $ref_class->newInstance();
+		if ( ! $ref_class->hasMethod( 'init' ) ) {
+			return $instance;
+		}
+
+		$init_method = $ref_class->getMethod( 'init' );
+		if ( ! $init_method->isPublic() || $init_method->isStatic() ) {
+			return $instance;
+		}
+
+		// phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		$init_args          = $init_method->getParameters();
+		$init_arg_instances = array_map(
+			function ( \ReflectionParameter $arg ) use ( $class_name, &$resolve_chain ) {
+				$arg_type = $arg->getType();
+				if ( ! ( $arg_type instanceof \ReflectionNamedType ) ) {
+					throw new ContainerException( "Error resolving '$class_name': argument '\${$arg->getName()}' doesn't have a type declaration." );
+				}
+				if ( $arg_type->isBuiltin() ) {
+					throw new ContainerException( "Error resolving '$class_name': argument '\${$arg->getName()}' is not of a class type." );
+				}
+				if ( $arg->isPassedByReference() ) {
+					throw new ContainerException( "Error resolving '$class_name': argument '\${$arg->getName()}' is passed by reference." );
+				}
+				return $this->get_core( $arg_type->getName(), $resolve_chain );
+			},
+			$init_args
+		);
+
+		// phpcs:enable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		$init_method->invoke( $instance, ...$init_arg_instances );
+
+		return $instance;
+	}
+
+	/**
+	 * Tells if the 'get' method can be used to resolve a given class.
+	 *
+	 * @param string $class_name The class name.
+	 * @return bool True if the class with the supplied name can be resolved with 'get'.
+	 */
+	public function has( string $class_name ): bool {
+		$class_name = trim( $class_name, '\\' );
+		return $this->is_class_allowed( $class_name ) || isset( $this->resolved_cache[ $class_name ] );
+	}
+
+	/**
+	 * Checks to see whether a class is allowed to be registered.
+	 *
+	 * @param string $class_name The class to check.
+	 *
+	 * @return bool True if the class is allowed to be registered, false otherwise.
+	 */
+	protected function is_class_allowed( string $class_name ): bool {
+		return StringUtil::starts_with( $class_name, self::WOOCOMMERCE_NAMESPACE, false );
+	}
+
+	/**
+	 * Tells if this class should be used as the core WooCommerce dependency injection container (or if the old ExtendedContainer should be used instead).
+	 *
+	 * By default, this returns true, to have it return false you can:
+	 *
+	 * 1. Define the WOOCOMMERCE_USE_OLD_DI_CONTAINER constant with a value of true; or
+	 * 2. Hook on the 'woocommerce_use_old_di_container' filter and have it return false (it receives the value of WOOCOMMERCE_USE_OLD_DI_CONTAINER, or false if the constant doesn't exist).
```

Comment:
🤔 Would it be worth adding something to the System Status Report showing whether the old DI container is being used? And/or adding something to WC Tracker? That way we could get a sense for how disruptive it might be to remove the old DI container in versionn 10.0.

Reply:
This makes sense, but I'd add this in a separate PR, after we discuss the approach to follow.